# pman_inversions

## Scripts associated with Harringmeyer &amp; Hoekstra, 2022

centromeres.sh: script for blasting deer mouse centromere satellite sequence to genome

dotplots.sh: script for performing nucmer alignments between two contigs (or contig versus itself) to make dotplots

flye.sh: script for running flye to create de novo genome assemblies from PacBio long-read sequencing data

fst.py: script for computing Fst between inversion and standard haplotypes using scikit-allel

genotyping.py: script for genotyping inversions from whole-genome resequencing data (vcfs) with PCA using scikit-allel

ld.sh: script for computing LD from vcfs for chromosome of interest

lostruct.r: script for performing local PCA using lostruct

mutational_load.r: script for computing mutational load in inversion or standard haplotypes with PopGenome

pca_heterozygosity.py: script for computing heterozygosity by sample and PCA for regions of interest using scikit-allel

raxml.sh: pipeline for creating maximum-likelihood trees with RAxML

sedef.sh: script for running SEDEF to identify segmental duplications

selection_simulations1.sh: script for performing forward-genetic simulations in SLiM to characterize selection on an inversion in the forest-prairie populations; inversion introduced into forest population

selection_simulations2.sh: script for performing forward-genetic simulations in SLiM to characterize selection on an inversion in the forest-prairie populations; inversion introduced into prairie population
